src/state/download/actions.js

```diff
@@ -49,59 +58,77 @@ export const removedSpecies = speciesName => {
   };
 };
 
-export const addedExperiment = experiment => {
+/**
+ * Takes an array of experiments and adds to users dataset via endpoint
+ * @param {array} experiments
+ */
+export const addExperiment = experiments => {
   return async (dispatch, getState) => {
-    let response, bodyData;
-
-    const dataSetId = getState().download.dataSetId;
-
-    if (!dataSetId) {
-      bodyData = {
-        data: {
-          [experiment.accession_code]: experiment.samples
-        }
-      };
-      response = await asyncFetch('/dataset/create/', {
-        method: 'POST',
-        headers: {
-          'content-type': 'application/json'
-        },
-        body: JSON.stringify(bodyData)
-      });
-
-      const { id } = response;
-      localStorage.setItem('dataSetId', id);
-    } else {
-      const prevDataSet = getState().download.dataSet;
-      const formattedDataSet = _formatDataSetObj(prevDataSet);
-      bodyData = {
-        data: {
-          ...formattedDataSet,
-          [experiment.accession_code]: experiment.samples
-        }
-      };
-
-      response = await asyncFetch(`/dataset/${dataSetId}/`, {
-        method: 'PUT',
-        headers: {
-          'content-type': 'application/json'
-        },
-        body: JSON.stringify(bodyData)
-      });
-    }
-
-    const { data } = response;
-
-    dispatch(fetchDownloadData(data));
     dispatch({
-      type: 'DOWNLOAD_EXPERIMENT_ADDED',
+      type: 'DOWNLOAD_ADD_EXPERIMENT',
       data: {
-        dataSetId
+        experiments
       }
     });
+    const dataSetId = getState().download.dataSetId;
+    const prevDataSet = getState().download.dataSet;
+    const formattedDataSet = prevDataSet;
+    const newExperiments = experiments.reduce((result, experiment) => {
+      if (experiment.samples.length)
+        result[experiment.accession_code] = experiment.samples;
+      return result;
+    }, {});
+    const bodyData = {
+      data: {
+        ...formattedDataSet,
+        ...newExperiments
+      }
+    };
+
+    try {
+      let response;
+      if (!dataSetId) {
+        response = await asyncFetch('/dataset/create/', {
+          method: 'POST',
+          headers: {
+            'content-type': 'application/json'
+          },
+          body: JSON.stringify(bodyData)
+        });
+
+        const { id } = response;
+        localStorage.setItem('dataSetId', id);
+      } else {
+        response = await asyncFetch(`/dataset/${dataSetId}/`, {
+          method: 'PUT',
+          headers: {
+            'content-type': 'application/json'
+          },
+          body: JSON.stringify(bodyData)
```

Comment:
Yes, currently the backend requires the entire dataset object to be passed in. I'm guessing this was to avoid separate endpoints for adding and removing experiments and samples?